code/localization/localize.cpp

```diff
@@ -724,199 +717,318 @@ void lcl_fred_replace_stuff(SCP_string &text)
-			*id = -2;
+		// the next non-whitespace char should be a (
+		ignore_white_space(&ch);
+		if (*ch != '(')
+			break;
+		ch++;
 
-		return false;
-	}
+		// the next should be a quote
+		ignore_white_space(&ch);
+		if (*ch != '\"')
+			break;
+		ch++;
 
-	// otherwise, check to see if it's an XSTR() tag
-	if (strnicmp(in, "XSTR", 4) != 0) {
-		// NOT an XSTR() tag
-		if (str_len > max_len && !Lcl_unexpected_tstring_check)
-			error_display(0, "Token too long: [%s].  Length = " SIZE_T_ARG ".  Max is " SIZE_T_ARG ".\n", in, str_len, max_len);
+		// now we have the start of the string
+		auto str_start = ch;
 
-		strncpy(out, in, max_len);
+		// find the end of the string
+		ch = strchr(ch, '"');
+		if (ch == nullptr)
+			break;
 
-		if (id != NULL)
-			*id = -2;
+		// now we have the end of the string (past the last character in it)
+		auto str_end = ch;
+		ch++;	// skip the quote
 
-		return false;
-	}
+		// the next non-whitespace char should be a ,
+		ignore_white_space(&ch);
+		if (*ch != ',')
+			break;
+		ch++;
 
-	// at this point we _know_ its an XSTR() tag, so split off the strings and id sections
-	if (!lcl_ext_get_text(in, text_str)) {
-		if (str_len > max_len && !Lcl_unexpected_tstring_check)
-			error_display(0, "Token too long: [%s].  Length = " SIZE_T_ARG ".  Max is " SIZE_T_ARG ".\n", in, str_len, max_len);
+		// check for number, being mindful of negative
+		ignore_white_space(&ch);
+		bool is_negative = false;
+		if (*ch == '-')
+		{
+			is_negative = true;
+			ch++;
+		}
+		if (!isdigit(*ch))
+			break;
 
-		strncpy(out, in, max_len);
+		// now we have the start of the id
+		auto id_start = ch;
 
-		if (id != NULL)
-			*id = -1;
+		// find all the digits
+		while (isdigit(*ch))
+			ch++;
 
-		return false;
-	}
-	if (!lcl_ext_get_id(in, &str_id)) {
-		if (str_len > max_len && !Lcl_unexpected_tstring_check)
-			error_display(0, "Token too long: [%s].  Length = " SIZE_T_ARG ".  Max is " SIZE_T_ARG ".\n", in, str_len, max_len);
+		// now we have the end of the id (past the last character in it)
+		auto id_end = ch;
 
-		strncpy(out, in, max_len);
+		// the next non-whitespace char should be a )
+		ignore_white_space(&ch);
+		if (*ch != ')')
+			break;
 
-		if (id != NULL)
-			*id = -1;
+		// if we got this far, we know we have a parseable XSTR of some sort
+		xstr_id = -1;
 
-		return false;
-	}
-	
-	// if the localization file is not open, or there's no entry, or we're not translating, return the original string
-	if ( !Xstr_inited || (str_id < 0) || (!use_default_translation && ((Lcl_current_lang == LCL_UNTRANSLATED) || (Lcl_current_lang == LCL_RETAIL_HYBRID))) ) {
-		if ( strlen(text_str) > max_len && !Lcl_unexpected_tstring_check )
-			error_display(0, "Token too long: [%s].  Length = " SIZE_T_ARG ".  Max is " SIZE_T_ARG ".\n", text_str, strlen(text_str), max_len);
+		//
+		// split off the strings and id sections
+		//
 
-		strncpy(out, text_str, max_len);
+		// check bounds
+		if (str_end - str_start > PARSE_BUF_SIZE - 1)
```

Comment:
Consider adding
```
int str_length = (int)(str_end - str_start);
```
above this line and then using `str_length` here and also below.

Reply:
Good idea.  Done.